**Az első 5 feladat ismétlés (többtáblás lekérdezések) a tanulmányi adatbázisból**

1. Készítsünk listát arról, hogy melyik óra melyik teremben van!
    
    1. Jelenítsük meg az óra azonosítóját és a terem nevét!
    2. A listát szűrjük a 3 karakter hosszú, \[100;199\] zárt intervallumba eső termekre! Feltesszük, hogy a 3 karakter hosszú terem nevek csak számjegyeket tartalmaznak.

In [6]:
SELECT o.ora_id,
        t.Terem
from orak o JOIN Termek t ON o.terem=t.Tkód
where t.Terem BETWEEN 100 and 199



Total execution time: 00:00:00.052

ora_id Terem 159 117 160 118 161 118 162 119 163 120 164 119 165 119 166 119 167 120 170 120 172 119 173 119 175 120 176 118 177 119 179 120 180 119 181 118 185 120

: Msg 245, Level 16, State 1, Line 1
Conversion failed when converting the nvarchar value 'VP 203.' to data type int.

2. <span style="font-weight: 600; color: rgb(36, 41, 47); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 16px;">Hány óra van összesen az egyes napokon?</span>

1. Jelenítsük meg a napok nevét és az órák számát az adott napon!
2. Az oszlopokat nevezzük el értelemszerűen!
3. A listát rendezzük óraszám szerint növekvő sorrendbe!

In [9]:
select n.nap,
        count(*) as 'Órák száma'
from Orak o JOIN Napok n ON o.nap=n.nap_id
group by n.nap
order by count(*) DESC


(5 rows affected)

Total execution time: 00:00:00.013

nap Órák száma Hétfő 23 Csütörtök 19 Kedd 13 Péntek 6 Szerda 6

3. <span style="font-weight: 600; color: rgb(36, 41, 47); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 16px;">Készítsünk listát arról, hogy melyik oktatónak hány órája van az Órák táblában!</span>

1. A lista jelenítse meg az oktatók nevét, és az órák számát!
2. Az oszlopokat nevezzük el értelemszerűen!
3. A listából hagyjuk ki az óraadókat!

4. <span style="color: rgb(36, 41, 47); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 16px;">Van-e olyan tantárgy, amelyből nincs óra?</span>

1. Jelentsük meg a tantárgy nevét és a tantárgyfelelős nevét!

5. <span style="background-color: rgb(255, 255, 255); color: rgb(36, 41, 47); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 16px;">Hány óra van összesen a 120-as teremben?</span>

**A 6-10 feladatokat halmazműveletek segítségével oldjuk meg, szintén a tanulmányi adatbázison**

6. <span style="background-color:rgb(255, 255, 255);color:rgb(36, 41, 47);font-family:-apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;;font-size:16px;font-weight:600;">Jelenítsük meg azon tanárok azonosítóját és nevét, akik hétfői vagy keddi napokon tanítanak!</span>

In [10]:
select okt.oktato_id,
        okt.nev
from Orak o JOIN Oktatok okt ON o.tanar=okt.oktato_id
            JOIN Napok nap ON o.nap=nap.nap_id
where nap.nap='Hétfő'

UNION

select okt.oktato_id,
        okt.nev
from Orak o JOIN Oktatok okt ON o.tanar=okt.oktato_id
            JOIN Napok nap ON o.nap=nap.nap_id
where nap.nap='Kedd'

(17 rows affected)

Total execution time: 00:00:00.045

oktato_id nev 1 Kovács László 2 Szentágotai Istvás 3 Szerb László 4 Ujfalussi István 5 Balkán András 6 Szép llonka 7 Kiss János 8 Pelikán József 9 Farkasos Piroska 10 Jogrerős Pál 12 Varga Kata 14 Szénási ászló 18 Nógrádi Kata 20 Kovács András 104 Ludas Mátyás 108 Róbert Gida 200 Aranyi Mária

7. <span style="background-color: rgb(255, 255, 255); color: rgb(36, 41, 47); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 16px; font-weight: 600;">Melyek azok a tantárgyak, amelyek esetén van óra a 8.00-kor kezdődő sávban, de nincs a 12.30-kor kezdődő sávban?</span>

In [12]:
select tan.Tantárgynév
from Orak o JOIN Savok sav On o.Sav=sav.Sav_id
            JOIN Tantárgyak tan ON o.targy=tan.Tkód
where sav.Időpont='8.00'

EXCEPT

select tan.Tantárgynév
from Orak o JOIN Savok sav On o.Sav=sav.Sav_id
            JOIN Tantárgyak tan ON o.targy=tan.Tkód
where sav.Időpont='12.30'

(3 rows affected)

Total execution time: 00:00:00.042

Tantárgynév Az ürgeöntés hydrológiai alapjai II. Fejezetek az ürgék jogaiból Ürgevitel

8. <span style="background-color: rgb(255, 255, 255); color: rgb(36, 41, 47); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 16px; font-weight: 600;">Listázzuk azokat a termeket, amelyekben oktat Kovács László és van olyan tanóra a teremben, ahol a tantárgy nevében nem szerepel az 'ürge' szó!</span>

In [13]:
select okt.nev,
        ter.Terem
from Orak o JOIN Oktatok okt ON o.tanar=okt.oktato_id
            JOIN Tantárgyak tan ON o.targy=tan.Tkód
            JOIN Termek ter ON o.terem=ter.Tkód
where okt.nev='Kovács László'

EXCEPT 

select okt.nev,
        ter.Terem
from Orak o JOIN Oktatok okt ON o.tanar=okt.oktato_id
            JOIN Tantárgyak tan ON o.targy=tan.Tkód
            JOIN Termek ter ON o.terem=ter.Tkód
where tan.Tantárgynév='%ürge%'

(4 rows affected)

Total execution time: 00:00:00.064

nev Terem Kovács László 116 Kovács László 119 Kovács László S.Asor.S3 Kovács László VP 203.

9. <span style="font-weight: 600; color: rgb(36, 41, 47); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 16px;">Készítsünk listát arról, hogy melyik tanárnak hány órája van a (munka) hét elején (hétfőn), hét közben (kedd, szerda és csütörtök), illetve a hét végén (péntek)!</span>

1. Jelenítsük meg a tanár azonosítóját, nevét, az időszakot (hét eleje - hét közben - hét vége), valamint az óraszámot!
2. A három megjelölt időszakot külön kérdezzük le, majd fűzzük össze a lekérdezések eredményeit!
3. Rendezzük a listát az oktató neve, azon belül időszak szerint!

10. <span style="background-color: rgb(255, 255, 255); color: rgb(36, 41, 47); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 16px; font-weight: 600;">Melyik napokon nincs órája Pelikán Józsefnek?</span>

In [14]:
select *
from Oktatok okts JOIN Orak o ON okt.oktato_id=o.tanar
where o.ora_id is NULL

(5 rows affected)

Total execution time: 00:00:00.048

oktato_id nev titulus beosztás status ora_id tanar nap Sav targy terem 11 Józsa Gábor Dr. 4 1 NULL NULL NULL NULL NULL NULL 19 Rácz András NULL 1 1 NULL NULL NULL NULL NULL NULL 105 Kiss István NULL 1 2 NULL NULL NULL NULL NULL NULL 107 Mészáros Kriszta NULL 1 2 NULL NULL NULL NULL NULL NULL 302 Kovacs Istvan Dr. 5 1 NULL NULL NULL NULL NULL NULL

**A 11-15 feladatok gyakorlásra szolgálnak a közelgő zh-ra :-)**

11. _Készítsünk listát az oktatók adatairól!_

1. Az oktatók titulusa és neve összefűzve jelenjen meg, közöttük szóközzel!
2. Ha valakinek nincs titulusa, akkor csak a neve jelenjen meg, szóköz nélkül!
3. A listát rendezzük a beosztás kódja, azon belül a státusz kódja szerint növekvő sorrendbe!
4. Az oszlopokat nevezzük el értelemszerűen!

In [18]:
select 
    case   
        when titulus is null then nev
        else titulus+' '+nev
    end as 'nev vagymi'
from Oktatok
order by titulus, [status]

(25 rows affected)

Total execution time: 00:00:00.038

nev vagymi Szerb László Kiss János Szénási ászló Nógrádi Kata Rácz András Kovács András Ludas Mátyás Róbert Gida Vitéz János Aranyi Mária Szabó Andrera Kiss István Mészáros Kriszta Pelikán József Farkasos Piroska Dr. Szép llonka Dr. Józsa Gábor Dr. Varga Kata Dr. Juhász András Dr. Kovacs Istvan Dr. Balkán András Dr. Kovács László Dr. Szentágotai Istvás Prof, tanszékvezető Ujfalussi István Prof. Jogrerős Pál

<span style="color: #09885a;">12</span>. Készítsünk listát, amely megmutatja, hogy összesen hány tantárgy van kreditenkénti csoportosításban!

  

A listából hagyjuk ki azokat a tételeket, ahol ez a darabszám <span style="color: #09885a;">1</span>!

A Folyami gáttan tantárgyat ne vegyük figyelembe!

A listát rendezzük darabszám szerint csökkenő sorrendbe!

Az oszlopokat nevezzük el értelemszerűen!

In [21]:
SELECT kredit AS 'Kreditek száma', 
       COUNT(*) AS 'DB'
FROM Tantárgyak
WHERE Tantárgynév <> 'Folyami gáttan'
GROUP BY kredit
HAVING COUNT(*)<>1
ORDER BY DB DESC

(4 rows affected)

Total execution time: 00:00:00.009

Kreditek száma DB 2 3 3 3 4 2 1 2

13. _Készítsünk listát az oktatók nevéről, a beosztásuk alapján elvárt és a tényleges óraszámukról!_

1. Az oszlopok neve legyen 'Oktató neve', 'Elvárt óraszám' és 'Tényleges óraszám'
2. Szűrjük a listát azokra az oktatókra, ahol az elvárt és a tényleges óraszám azonos

In [29]:
select okt.nev,
    beo.elvart,
            count(*) as 'tényleges óra'
from Oktatok okt JOIN Orak o ON okt.oktato_id=o.tanar
            JOIN Beosztasok beo ON okt.beosztás=beo.beosztas_id
group by okt.nev, beo.elvart
having count(*)=beo.elvart


(4 rows affected)

Total execution time: 00:00:00.042

nev elvart tényleges óra Jogrerős Pál 2 2 Juhász András 2 2 Varga Kata 4 4 Pelikán József 6 6

14. _Készítsünk listát, amely megmutatja, hogy mely tantárgyból nincs óra egyáltalán!_

1. Csak a tantárgy neve jelenjen meg!

In [30]:
select *
FROM Orak o RIGHT JOIN Tantárgyak tan ON o.targy=tan.Tkód
where o.ora_id is NULL

(1 row affected)

Total execution time: 00:00:00.054

ora_id tanar nap Sav targy terem Tkód Tantárgynév Tantárgyfelelős Tantargykod Kredit NULL NULL NULL NULL NULL NULL 11 Folyami gáttan 12 NULL 2

15. _Hány órájuk van az óraadóknak napi bontásban?_

1. Csak a napok nevét és az óraszámot jelenítsük meg!
2. Rendezzük a listát óraszám szerint növekvő sorrendbe!